web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeQueryLogic.java

```diff
@@ -109,14 +116,23 @@ public void run() {
                             results.add(last);
                         }
                     } catch (InterruptedException e) {
-                        log.warn("QueryLogic thread interrupted", e);
+                        // if this was on purpose, then just log and the loop will naturally exit
+                        if (interrupted) {
+                            log.warn("QueryLogic thread interrupted", e);
+                        }
+                        // otherwise something else unexpected happened
+                        else {
+                            throw new RuntimeException(e);
+                        }
                     }
                     resultCount++;
                 }
-                
+                success = true;
             } finally {
-                completionLatch.countDown();
-                log.trace("Finished thread: " + this.getName());
+                if (success) {
+                    completionLatch.countDown();
+                }
```

Comment:
I'm trying to convince myself that we only want to call `countDown()` on `success`. Docs seems to suggest that until the latch reaches zero, an `await()` call elsewhere will block. If we don't `countDown()` on failure it stands to reason that the counter will never reach zero and this something somewhere will fail to unblock. That said, I haven't completed an analysis to understand how this latch is truly used.

Reply:
The latches are need to know when a query logic has completed.  If not successful, then an exception should have been logged which will also break the loop in the CompositeQueryLogicResultsIterator loop in the hasNext call.